### PR TITLE
refactor(rxAutoSave): FRMW-398-convert to Quark Service.

### PR DESCRIPTION
**Pull request by 100stacks**
JIRA: https://jira.rax.io/browse/FRMW-398

- [x] Dev LGTM
- [x] Dev LGTM
- [x] Design LGTM

---
Refactor `rxMisc.rxAutoSave` to Quark Service.

### CODE REVIEW

**Comment by 100stacks** [comment]
## `rxAutoSave` Quark Service

### Demo Page
screen shot 2016-01-12 at 10 11 49 am

### API Docs
screen shot 2016-01-12 at 10 12 28 am

File: src/components/rxMisc/docs/rxMisc.midway.js [hunk 1]
```diff
@@ -1,11 +1,5 @@
-var _ = require('lodash');
 var Page = require('astrolabe').Page;
 
-// "wait" for autosave to clear -- function passed to `browser.wait`
```
Comment (100stacks) [inline src/components/rxMisc/docs/rxMisc.midway.js]: Breaking out `rxAutoSave` midway logic.

File: src/components/rxMisc/docs/rxMisc.midway.js [hunk 2]
```diff
@@ -22,19 +16,6 @@ var autoSaving = Page.create({
 
     sensitiveData: encore.rxForm.textField.generateAccessor(element(by.model('formData.sensitive'))),
 
-    clearAutoSave: {
```
Comment (100stacks) [inline src/components/rxMisc/docs/rxMisc.midway.js]: Breaking out `rxAutoSave` midway logic.

File: src/components/rxMisc/docs/rxMisc.midway.js [hunk 3]
```diff
@@ -42,70 +23,6 @@ describe('rxMisc', function () {
         demoPage.go('#/components/rxMisc');
     });
 
-    describe('rxAutoSave', function () {
```
Comment (100stacks) [inline src/components/rxMisc/docs/rxMisc.midway.js]: Breaking out `rxAutoSave` midway logic.

File: src/components/rxMisc/rxMisc.module.js [hunk 4]
```diff
@@ -4,10 +4,7 @@
  * @description
  * # rxMisc Component
  *
- * A module for shared functionality across framework components.
- *
- * ## Services
- * * {@link rxMisc.service:rxAutoSave rxAutoSave}
+ * __NOTE:__ Marked for DEPRECATION.  All `rxMisc` components refactored to `Quarks`.
```
Comment (100stacks) [inline src/components/rxMisc/rxMisc.module.js]: Update description to indicate `rxMisc` component is marked for deprecation.  Maintaining `rxMisc.module.js` as the `convenience functions` page objects are still located here.

File: src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Once you've tried that, you can clear the stored data with either of the "Clear" buttons below. Clicking those
+        will <em>not</em> erase the form, but will erase the rxAutoSave stored data. If you navigate away after clicking
+        one of the buttons, and then return here, the form will not be auto-populated, because rxAutoSave will have no data for it!
+    </p>
+
+    <p>
+        Finally, we have set a Time-To-Live of 24 hours for this form, and explicitly told it not to save the "Sensitive data" field
+        into storage. If you enter data here, and come back &gt;24 hours later, the form will not be auto-populated, as the data will
+        have expired.
+    </p>
```
Comment (CITguy) [inline src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html]: I'd pull the `h3` and descriptive paragraphs out into the `docs/rxAutoSave.docs.html` to keep the source for the example as small as possible.

File: src/quarks/quarks.module.js [hunk 6]
```diff
@@ -30,10 +30,14 @@
...
+ * * {@link quarks.service:ErrorFormatter ErrorFormatter}
+ * * {@link quarks.service:rxFeedbackSvc rxFeedbackSvc}
+ * * {@link quarks.service:rxNotify rxNotify}
+ * * {@link quarks.service:rxAutoSave rxAutoSave}
+ * * {@link quarks.service:rxDOMHelper rxDOMHelper}
+ * * {@link quarks.service:rxScreenshotSvc rxScreenshotSvc}
  */
 angular.module('encore.ui.quarks', [
     'ngResource',
+    'debounce',
```
Comment (100stacks) [inline src/quarks/quarks.module.js]: `debounce` dependency pulled in to support `rxAutoSave` Quark Service.

File: src/quarks/rxAutoSave/docs/rxAutoSave.midway.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            description: ['Using snow as cover, leap from your hiding spot to',
+                          'surprise your enemy in a display of tactical brilliance.',
+                          'In order to prevent your outfit from soaking',
+                          '(later risking hypothermia) you must wear the least',
+                          'amount of clothes possible to stay dry through the night.'].join(' '),
+            sensitiveData: 'Only Jay Parlar knows where the Candian Armed Forces lie in wait!'
+        };
+
+        var leavePage = function () {
+            demoPage.go('#/components/rxPaginate');
```
Comment (CITguy) [inline src/quarks/rxAutoSave/docs/rxAutoSave.midway.js]: I'd change this to something less likely to break as we migrate Quarks/Atoms/Molecules. Try `#/quarks/titelize`, instead (considering the location shouldn't move any time soon).

File: src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js [hunk 8]
```diff
@@ -27,9 +27,8 @@ describe('rxAutoSave', function () {
     };
 
     beforeEach(function () {
-        module('encore.ui.rxMisc');
-        module('encore.ui.rxLocalStorage');
         module('encore.ui.quarks');
+        module('encore.ui.rxLocalStorage');
```
Comment (CITguy) [inline src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js]: So, I'm thinking that this PR may benefit from LocalStorage migrating first (FRMW-636).

**Comment by CITguy** [comment]
Added DO NOT MERGE: should be completed after `rxLocalStorage.LocalStorage` is migrated.

**Comment by 100stacks** [comment]
Adding reference to `rxLocalStorage` migration https://github.com/rackerlabs/encore-ui/pull/1494.

**Comment by 100stacks** [comment]
@CITguy, comments implemented. :+1:

File: src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html [hunk 9]
```diff
@@ -0,0 +1,72 @@
+<div ng-controller="rxAutoSaveSimpleCtrl">
```
Comment (CITguy) [inline src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html]: Please correct indentation of file to be consistent with 2 spaces per indentation level.

File: src/quarks/rxAutoSave/rxAutoSave.meta.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+    "displayName": "rxAutoSave",
+    "stability": "unstable",
+    "description": "Provides a way to store values in a form for later use.",
```
Comment (CITguy) [inline src/quarks/rxAutoSave/rxAutoSave.meta.json]: I'd update to `Service to help automatically save/load form data.`

File: src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js (hunk 8 above)
Comment (CITguy) [inline src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js]: This line will be unnecessary once #1494 is merged.

**Comment by 100stacks** [comment]
@CITguy, updated description and indentation. :+1:

File: src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js (hunk 8 above)
Comment (100stacks) [inline src/quarks/rxAutoSave/scripts/rxAutoSave.spec.js]: @CITguy, ok I'll remove that line and rebase after https://github.com/rackerlabs/encore-ui/pull/1494 is merged.

File: src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+          disable="status.disable">
+        </rx-button>
+
+        <rx-button
+          ng-click="deferredClear()"
+          toggle="status.deferredLoading"
+          default-msg="Clear rxAutoSave by resolving a promise"
+          toggle-msg="Clearing rxAutoSave data"
+          disable="status.deferredDisable">
+      </rx-button>
```
Comment (CITguy) [inline src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html]: Indent line one more level.

**Comment by 100stacks** [comment]
@CITguy, indented the `<rx-button>` blocks an additional level. :+1:

File: src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html [hunk 12]
```diff
@@ -0,0 +1,72 @@
+<div ng-controller="rxAutoSaveSimpleCtrl">
+  <form name="demoForm" rx-form>
+    <rx-form-section stacked>
+      <rx-field>
+        <rx-field-name>A checkbox field!:</rx-field-name>
```
Comment (glynnis) [inline src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html]: I think we can lose the `!` here -- looks funny next to a colon. Not sure if this was an error or not :sweat_smile:

**Comment by glynnis** [comment]
Other than the `!` this looks good from a design standpoint :+1:

File: src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            disable="status.disable">
+        </rx-button>
+
+        <rx-button
+            ng-click="deferredClear()"
+            toggle="status.deferredLoading"
+            default-msg="Clear rxAutoSave by resolving a promise"
+            toggle-msg="Clearing rxAutoSave data"
+            disable="status.deferredDisable">
+      </rx-button>
```
Comment (CITguy) [inline src/quarks/rxAutoSave/docs/examples/rxAutoSave.simple.html]: Please correct indentation for closing `</rx-button>`

**Comment by 100stacks** [comment]
@glynnis, that was a happy checkbox`!` :white_check_mark: :+1:

**Comment by CITguy** [comment]
LGTM

**Comment by 100stacks** [comment]
@CITguy, okay fixed the stray `<rx-button>` :+1:

**Comment by CITguy** [comment]
Removed "DO NOT MERGE" label. #1494 has been merged

Please rebase.

**Comment by 100stacks** [comment]
@CITguy, updated and rebased :+1:

File: src/components/rxMisc/docs/rxMisc.midway.js [hunk 14]
```diff
@@ -1,11 +1,5 @@
-var _ = require('lodash');
 var Page = require('astrolabe').Page;
 
-// "wait" for autosave to clear -- function passed to `browser.wait`
-var forAutoSaveToClear = function () {
-    return encore.rxNotify.all.exists('rxAutoSave data has been cleared!');
-};
-
 // anonymous page object
 var autoSaving = Page.create({
```
Comment (CITguy) [inline src/components/rxMisc/docs/rxMisc.midway.js]: Is this anonymous page object still needed?

**Comment by 100stacks** [comment]
Updated the API Docs to reference the `EncoreUI` provided `rxLocalStorage` service.  We are deprecating `LocalStorage` service to reduce confusion/conflict with the browser's `localStorage` object.

**Pull request merged**